pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/adminoperation/ParticipantInitService.java

```diff
@@ -0,0 +1,201 @@
+            if (probandData == null) {
+                return new UpdateLog(ddpParticipantId, UpdateLog.UpdateStatus.ERROR.name(),
+                        "Participant has family ID in ES but no RGP data");
+            }
+            return new UpdateLog(ddpParticipantId, UpdateLog.UpdateStatus.NOT_UPDATED.name());
+        }
+
+        // participant has no family ID in ES
+        try {
+            // ptp does not have a family ID in ES, but does have RGP participant data
+            if (probandData != null) {
+                return updateESFamilyId(probandData, ddpParticipantId, ddpInstance.getParticipantIndexES(), isDryRun);
+            }
+            return createDefaultData(ddpParticipantId, esParticipant, ddpInstance, familyIdProvider, isDryRun);
+        } catch (ESMissingParticipantDataException e) {
+            // no profile
+            return new UpdateLog(ddpParticipantId, UpdateLog.UpdateStatus.NO_PARTICIPANT_DATA.name());
+        } catch (Exception e) {
+            String msg = String.format("Exception in ParticipantInitService.run for participant %s: %s",
+                    ddpParticipantId, e);
+            // many of these exceptions will require investigation, but conservatively we will just log
+            // at error level for those that are definitely concerning
+            if (e instanceof DsmInternalError) {
+                log.error(msg);
+                e.printStackTrace();
+            } else {
+                log.warn(msg);
+            }
+            return new UpdateLog(ddpParticipantId, UpdateLog.UpdateStatus.ERROR.name(), e.toString());
+        }
+    }
+
+    protected static ParticipantData getProbandRgpData(List<ParticipantData> ptpData, UpdateLog updateLog) {
+        List<ParticipantData> probandData = ptpData.stream()
+                .filter(data -> MEMBER_TYPE_SELF.equals(data.getDataMap().get(MEMBER_TYPE)))
+                .collect(Collectors.toList());
+        if (probandData.isEmpty()) {
+            updateLog.setError("Participant has RGP data but no proband record");
+            return null;
+        }
+        if (probandData.size() > 1) {
+            updateLog.setError(String.format("Participant has %d RGP proband records", probandData.size()));
+            return null;
+        }
+
+        ParticipantData participantData = probandData.get(0);
+        String familyId = participantData.getDataMap().get(FAMILY_ID);
+        if (StringUtils.isBlank(familyId)) {
+            updateLog.setError("Participant has proband RGP data that does not contain a family ID");
+        }
+        return participantData;
+    }
+
+    protected static UpdateLog createDefaultData(String ddpParticipantId, ElasticSearchParticipantDto esParticipant,
+                                                 DDPInstance ddpInstance, FamilyIdProvider familyIdProvider,
+                                                 boolean isDryRun) {
+        if (isDryRun) {
```

Comment:
Is dryrun just doing some validations and not actually doing any updates in DB and then rolling back ?
I know we cannot do any updates to elastic. 
Trying to understand dryrun purpose

Reply:
Dry run just looks the data in ES and the DB, and logs what action to expect in when not doing a dry run. For instance, it will say UPDATED or NOT_UPDATED or ERROR, etc. This provides a full log of what to expect when this is run for real.
In non-dry-run mode (real changes), this will update the DB and/or ES.